ccm_web/client/src/utils/ThirdPartyGradebookProcessor.tsx

```diff
@@ -0,0 +1,142 @@
+import { CSVRecord } from '../utils/FileParserWrapper'
+import { InvalidationType } from '../models/models'
+
+export interface GradebookUploadRecord extends CSVRecord {
+  'SIS Login ID': string
+}
+
+export const REQUIRED_LOGIN_ID_HEADER = 'SIS Login ID'
+const OTHER_REQUIRED_HEADERS = ['Student Name', 'Student ID', 'SIS User ID', 'Section']
+export const REQUIRED_ORDERED_HEADERS = [
+  ...OTHER_REQUIRED_HEADERS.slice(0, 3), REQUIRED_LOGIN_ID_HEADER, OTHER_REQUIRED_HEADERS[3]
+]
+export const POINTS_POS_TEXT = 'Points Possible'
+
+export const isGradebookUploadRecord = (record: CSVRecord): record is GradebookUploadRecord => {
+  return typeof record[REQUIRED_LOGIN_ID_HEADER] === 'string'
+}
+
+export interface GradebookInvalidation {
+  message: string
+  type: InvalidationType
+}
+
+interface ProcessResultSuccess {
+  valid: true
+  processedRecords: GradebookUploadRecord[]
+  assignmentHeader: string
+  invalidations: GradebookInvalidation[]
+}
+
+interface ProcessResultFailure {
+  valid: false
+  invalidations: GradebookInvalidation[]
+}
+
+export type ProcessResult = ProcessResultSuccess | ProcessResultFailure
+
+export default class ThirdPartyGradebookProcessor {
+  studentLoginIds: string[]
+
+  constructor (studentLoginIds: string[]) {
+    this.studentLoginIds = studentLoginIds
+  }
+
+  static detectPointsPossible (firstRecord: GradebookUploadRecord): GradebookInvalidation | undefined {
+    if (!Object.values(firstRecord).includes(POINTS_POS_TEXT)) {
+      return {
+        message: `The file you uploaded is missing a ${POINTS_POS_TEXT} row.`,
+        type: InvalidationType.Error
+      }
+    }
+  }
+
+  static detectAssignment (oneRecord: GradebookUploadRecord): [string, undefined] | [undefined, GradebookInvalidation] {
+    const otherKeys = Object.keys(oneRecord).filter(k => !REQUIRED_ORDERED_HEADERS.includes(k))
+    if (otherKeys.length === 1) return [otherKeys[0], undefined]
+    let message: string | undefined
+    if (otherKeys.length === 0) {
+      message = 'No assignment column was found.'
+    } else {
+      message = 'Multiple assignment columns were found; only one assignment column at a time is supported.'
+    }
+    return [undefined, { message, type: InvalidationType.Error }]
+  }
```

Comment:
@pushyamig, do I need to validate the assignment name for length? `0 < value <= 255`?

Reply:
Canvas currently is throwing an  error `{"errors":{"title":[{"attribute":"title","type":"The title cannot be longer than 255 characters","message":"The title cannot be longer than 255 characters"}]}}` during the gradebook import, SO it will good if we have a safety net around it. 

Speaking practically this case will not be encounter.

Reply:
Alright, I handled this here: https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/223/commits/36fa938172c19b76bec605f747fc51d607fcdeb5

It's probably not an optimal implementation, but I'd rather not try to refactor all the validation code across the project as part of this feature PR.

Reply:
>  but I'd rather not try to refactor all the validation code across the project as part of this feature PR.

Valid!